### PR TITLE
Add Network Server documentation

### PR DESCRIPTION
**Pull request by rvolosatovs**
#### Summary

References #783 

#### Changes

- Add Network Server documentation

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, database and configuration, according to the stability commitments in `README.md`.
- [x] Testing: The changes are covered with unit tests. The changes are tested manually as well.
- [x] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: doc/content/components/network-server.md [hunk 1]
```diff
@@ -7,3 +7,51 @@ weight: 3
 The Network Server handles the LoRaWAN network layer, including MAC commands, regional parameters and adaptive data rate (ADR).
 
 <!--more-->
+
+## Device Management
+
+Network Servers expose `Get`, `Set` and `Delete` gRPC methods for device management. REST analogues for these methods exist as well. Typical clients of these RPC methods are [Console]({{< ref "/components/console.md" >}}) and [CLI]({{< ref "/components/cli.md" >}}).
```
Comment (htdvisser) [inline doc/content/components/network-server.md]: I think it's sufficient to say that it exposes the `NsEndDeviceRegistry` service for managing end devices, and refer to `/reference/api/end_device#the-nsenddeviceregistry-service` for details.

I think it's more useful to explain what the NS stores about end devices than the names RPCs it exposes.

File: doc/content/components/network-server.md [hunk 2]
```diff
@@ -7,3 +7,51 @@ weight: 3
...
+
+In case the device has an active session and it is in Class B or C, a downlink task may added to internal downlink task queue following the device update.
```
Comment (htdvisser) [inline doc/content/components/network-server.md]: I don't know if we should go into implementation details here. Also we're using the term "downlink task queue" that we haven't explained yet.

File: doc/content/components/network-server.md [hunk 3]
```diff
@@ -7,3 +7,51 @@ weight: 3
...
+
+If Network Server successfully constructed a `DownlinkMessage`, it is passed to Gateway Server to schedule via `ScheduleDownlink`.
+
+## Uplink Handling
+
+Network Servers expose `HandleUplink` gRPC method, which is used to receive device uplinks. Typical clients of this RPC method are [Gateway Servers]({{< ref "/components/gateway-server.md" >}}).
+
+Network Servers process the uplinks received and handle accordingly. The first step is matching the uplink to a device. In case an uplink cannot be matched to a device stored in Network Server, it is dropped.
+
+- If Join-Request is received:
```
Comment (htdvisser) [inline doc/content/components/network-server.md]: Avoid bullet lists in documentation

**Comment by rvolosatovs** [comment]
If you look at Gateway Server documentation, there are quite some implementation details provided there: https://github.com/TheThingsNetwork/lorawan-stack/blob/master/doc/content/components/gateway-server.md#communication-with-network-server

**Comment by htdvisser** [comment]
If we decide that this section is about giving users/operators an overview of what each component does, then I think it needs to be black-box, so that text wouldn't belong here either indeed.

I think the purpose of this section should be:

- What are the responsibilities of a component?
- How do users interact with it?
- What is its interaction with gateways?
- What is its interaction with end devices?
- What is its interaction with applications?
- What does it store?

But let's wait to hear what @johanstokking has to say about it.

**Comment by johanstokking** [comment]
@rvolosatovs I understand and like the consistency taken from the GS doc, but I must agree with @htdvisser that this is not something that we should put under the component description.

I think that we should describe components indeed from the outside and not the inner workings per se. I would include the high level interaction between other components, without diving in the gRPC details, but just the relations it has and what they are for.

File: doc/content/components/network-server.md (hunk 2 above)
Comment (johanstokking) [reply doc/content/components/network-server.md]: I think it's safe to say here that if you update device settings that may trigger a downlink message (like desired MAC parameters change), the NS schedules the downlink message.

File: doc/content/components/network-server.md [hunk 4]
```diff
@@ -7,3 +7,51 @@ weight: 3
...
+
+## Device Management
+
+Network Servers expose `Get`, `Set` and `Delete` gRPC methods for device management. REST analogues for these methods exist as well. Typical clients of these RPC methods are [Console]({{< ref "/components/console.md" >}}) and [CLI]({{< ref "/components/cli.md" >}}).
+
+In case the device has an active session and it is in Class B or C, a downlink task may added to internal downlink task queue following the device update.
+
+## Application Downlink Queue Management
+
+Network Servers expose `DownlinkQueueReplace`, `DownlinkQueuePush` and `DownlinkQueueList` gRPC methods for application downlink management. Typical clients of these RPC methods are [Application Servers]({{< ref "/components/application-server.md" >}}).
```
Comment (johanstokking) [inline doc/content/components/network-server.md]: Probably same as above; no need for rpc names, but a link to the API would be helpful, and the fact that NS exposes downlink queue replace, push and list operations to AS.

File: doc/content/components/network-server.md [hunk 5]
```diff
@@ -7,3 +7,51 @@ weight: 3
...
+
+In case the device has an active session and it is in Class B or C, a downlink task may added to internal downlink task queue following the application queue update.
```
Comment (johanstokking) [inline doc/content/components/network-server.md]: Same and onwards

File: doc/content/components/network-server.md [hunk 6]
```diff
@@ -7,3 +7,51 @@ weight: 3
...
+
+In case link is not active, but Network Server has application-specific `ApplicationUp` messages to send, those messages will be queued and sent once the link is established.
+
+## Downlink Scheduling
+
+Network Server maintains internal downlink task queue. Each downlink task has an execution time associated with it, which tasks are sorted by in ascending order. Whenever a downlink task is ready to execute, it is executed as soon as possible.
+
+As first step of scheduling a downlink, Network Server attempts to construct a `DownlinkMessage`.
+
+In case a pending session exists and Join-Accept is queued for the device, `DownlinkMessage` will contain the Join-Accept. Otherwise, Network Server attempts to genererate a data downlink in the active session, which `DownlinkMessage` will contain.
```
Comment (johanstokking) [inline doc/content/components/network-server.md]: join-accept no title case

File: doc/content/components/network-server.md (hunk 3 above)
Comment (johanstokking) [reply doc/content/components/network-server.md]: Join-request no title case

Comment (johanstokking) [reply doc/content/components/network-server.md]: Indeed use small headings here and a numbered list to show order

**Comment by johanstokking** [comment]
@rvolosatovs please request review when ready for review

**Comment by rvolosatovs** [comment]
I will

File: api/api.swagger.json [hunk 7]
```diff
@@ -3815,6 +3815,7 @@
     },
     "/ns/applications/{end_device_ids.application_ids.application_id}/devices/{end_device_ids.device_id}": {
       "get": {
+        "summary": "GenerateDevAddr requests a device address assignment from the Network Server.",
```
Comment (rvolosatovs) [inline api/api.swagger.json]: Wonder why this was not generated before? :thinking: @htdvisser any idea?

Comment (htdvisser) [reply api/api.swagger.json]: Strange indeed. No idea.

File: doc/content/components/network-server.md [hunk 8]
```diff
@@ -7,3 +7,54 @@ weight: 3
 The Network Server handles the LoRaWAN network layer, including MAC commands, regional parameters and adaptive data rate (ADR).
 
 <!--more-->
+
+## Device Management
+
+Network Servers expose [NsEndDeviceRegistry]({{< ref "/reference/api/end_device#the-nsenddeviceregistry-service" >}}) service for end device management. Typical clients of this sevice are [Console]({{< ref "/components/console.md" >}}) and [CLI]({{< ref "/components/cli.md" >}}).
```
Comment (johanstokking) [inline doc/content/components/network-server.md]: Typo; `sevice`

And awesome that GitHub Actions picks this up as well

File: doc/content/components/network-server.md [hunk 9]
```diff
@@ -7,3 +7,54 @@ weight: 3
...
+
+Change of application downlink queue may trigger a downlink message.
+
+Once the link is established, Network Server will send application-specific uplink messages to the client via the link. There can be at most one active link per-appplication.
+
+In case link is not active, but Network Server has application-specific uplink messages to send, those messages will be queued and sent once the link is established.
+
+## Downlink Scheduling
+
+Network Server maintains internal downlink task queue. Each downlink task has an execution time associated with it, which tasks are sorted by in ascending order. Whenever a downlink task is ready to execute, it is executed as soon as possible.
```
Comment (johanstokking) [inline doc/content/components/network-server.md]: I find this too much behavior description and not what NS is, but that is my opinion. No need for change, but let's be aware of a sliding edge that we cannot exhaustively describe all our component's behavior, and nobody really cares

File: doc/content/components/network-server.md [hunk 10]
```diff
@@ -7,3 +7,54 @@ weight: 3
...
+
+## Uplink Handling
+
+Network Servers expose [GsNs]({{< ref "/reference/api/network_server#the-gsns-service" >}}) service for uplink handling. Typical clients of this service are [Gateway Servers]({{< ref "/components/gateway-server.md" >}}).
+
+Network Servers process the uplinks received and handle accordingly. The first step is matching the uplink to a device. In case an uplink cannot be matched to a device stored in Network Server, it is dropped.
+
+### Join-request
+
+If join-request is received:
```
Comment (johanstokking) [inline doc/content/components/network-server.md]: Empty line after

File: doc/content/components/network-server.md [hunk 11]
```diff
@@ -7,3 +7,54 @@ weight: 3
...
+1. Device is matched using the `DevEUI` and `JoinEUI` pair present in the join-request, which uniquely identifies the device.
+2. New `DevAddr` is assigned to the device and new MAC state is derived for the device.
+3. If [Join Server]({{< ref "/components/join-server.md" >}}) is present in the cluster, Network Server sends a join-request message to the cluster-local [Join Server]({{< ref "/components/join-server.md" >}}).
+4. If [Join Server]({{< ref "/components/join-server.md" >}}) is not present in the cluster or the device is not provisioned in the cluster-local [Join Server]({{< ref "/components/join-server.md" >}}), Network Server sends a join-request message to the [Join Server]({{< ref "/components/join-server.md" >}}) discovered via [interoperability configuration]({{< ref "/reference/interop-repository" >}}).
+5. If a [Join Server]({{< ref "/components/join-server.md" >}}) accepted the join-request, join-accept message may be enqueued for the device and application-specific uplink messages carrying relevant information about join-accept is sent to linked [Application Server]({{< ref "/components/application-server.md" >}}).
```
Comment (johanstokking) [inline doc/content/components/network-server.md]: I'm OK with the flow description, but imo this does not belong in the NS documentation. It would be a higher level description of how components interact on key LoRaWAN procedures

File: doc/content/components/network-server.md [hunk 12]
```diff
@@ -7,3 +7,54 @@ weight: 3
...
+
+### Data uplink
+If data uplink is received:
```
Comment (johanstokking) [inline doc/content/components/network-server.md]: Empty line after heading and after this line

File: doc/content/reference/api/network_server.md [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+title: "Network Server APIs"
+description: ""
+weight: 9
+---
+
+## The `Ns` service
+
+{{< proto/method service="Ns" method="GenerateDevAddr" >}}
+
+## The `AsNs` service
```
Comment (htdvisser) [inline doc/content/reference/api/network_server.md]: How is this relevant for end users? This is an internal service, right?

File: doc/content/reference/api/network_server.md [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+
+{{< proto/method service="AsNs" method="LinkApplication" >}}
+
+{{< proto/method service="AsNs" method="DownlinkQueueReplace" >}}
+
+{{< proto/method service="AsNs" method="DownlinkQueuePush" >}}
+
+{{< proto/method service="AsNs" method="DownlinkQueueList" >}}
+
+## The `GsNs` service
```
Comment (htdvisser) [inline doc/content/reference/api/network_server.md]: How is this relevant for end users? This is an internal service, right?

File: doc/content/components/network-server.md (hunk 9 above)
Comment (htdvisser) [inline doc/content/components/network-server.md]: "The Network Server maintains" (with article) or "Network Servers maintain" (plural)

File: doc/content/components/network-server.md [hunk 15]
```diff
@@ -7,3 +7,54 @@ weight: 3
...
+
+In case a pending session exists and join-accept is queued for the device, it is scheduled.
+
+### Data downlink
+
+In case a pending session does not exist or join-accept for it has already been sent, Network Server attempts to genererate and schedule data downlink in the active session.
+
+## Uplink Handling
+
+Network Servers expose [GsNs]({{< ref "/reference/api/network_server#the-gsns-service" >}}) service for uplink handling. Typical clients of this service are [Gateway Servers]({{< ref "/components/gateway-server.md" >}}).
```
Comment (htdvisser) [inline doc/content/components/network-server.md]: These gRPC services are internal API. IMO that doesn't belong in user documentation

**Comment by rvolosatovs** [comment]
As per your request, I added information about MIC checks, FCnt checks and MAC commands.
To what extent do you want that to be documented? We don't need to rewrite the LoRaWAN spec here, right?

**Comment by htdvisser** [comment]
In the component documentation I think it's sufficient to mention *that* it happens. In the future we can add a reference to give details on *how* it happens.

**Comment by rvolosatovs** [comment]
That's already done - please see https://github.com/TheThingsNetwork/lorawan-stack/pull/1630/files#diff-3c9ccd58cf3b37c27482477923206489R58

**Comment by htdvisser** [comment]
For the end device matching I think we can give some more details. The NS does a lookup of all registered end devices with a given DevAddr. Then it filters out "elegible" devices (based on FCnt, FPlan, ...?). If there are still multiple elegible devices, it calculates MICs for each device until it finds a matching MIC.

For MAC commands I think we should also mention that the NS schedules responses to MAC commands (such as LinkCheck) and generates MAC commands (such as LinkAdr).

For ADR in the component documentation I think it's sufficient to mention *that* it happens. I expect users to be interested in knowing which component does this. In the future we can add a reference to give details on *how* exactly the ADR algorithm works.

**Comment by rvolosatovs** [comment]
> For the end device matching I think we can give some more details. The NS does a lookup of all registered end devices with a given DevAddr. Then it filters out "elegible" devices (based on FCnt, FPlan, ...?). If there are still multiple elegible devices, it calculates MICs for each device until it finds a matching MIC.

this is already explained

**Comment by rvolosatovs** [comment]
Merging this
@htdvisser feel free to open a PR
